### ISSUE TITLE
Support request for private repositories on Docker Swarm

### ISSUE THREAD

**Issue by ziggekatten** [openfaas/faas-swarm#57]
## My actions before raising this issue
- [X ] Followed the [troubleshooting guide](https://docs.openfaas.com/deployment/troubleshooting/)
- [X ] Read/searched [the docs](https://docs.openfaas.com/)
- [X ] Searched [past issues](/issues)

So, deploying a function through either faas-cli returns success (202), and the function shows up in the UI with status not ready.

## Expected Behaviour

The function should be available to execute

## Current Behaviour

The function have status "not ready" in UI, and the images are never pulled to my faas-cluster.

There are no errors in any logs (nats, gateway, faas-swarn) and I do not know how to further investigate the issue. faas-cli login works toward registry. so does build and push. 

faas-cli version 0.9.3 (tried 0.9.2 as well) Docker Swarm 19.03.2

command used: "faas-cli deploy -a -f .\myfunc.yml"

Response "Deployed. 202 Accepted."

WHere should I start looking?

## Steps to Reproduce (for bugs)

1.
2.
3.
4.

## Your Environment

* FaaS-CLI version ( Full output from: `faas-cli version` ): 0.9.3

* Docker version `docker version` (e.g. Docker 17.0.05 ): 19.03.2

* Are you using Docker Swarm or Kubernetes (FaaS-netes)? Swarm

* Operating System and version (e.g. Linux, Windows, MacOS): Ubuntu Linux

* Code example or link to GitHub repo or gist to reproduce problem:

* Other diagnostic information / logs from [troubleshooting guide](https://docs.openfaas.com/deployment/troubleshooting)

## Next steps

You may [join Slack](https://docs.openfaas.com/community) for community support.

**Comment by alexellis** [comment]
Have you run `docker login harbor.coreteam.nu` on your client computer at all?

**Comment by alexellis** [comment]
When you say `config.json` "is empty", can you share that "empty" file?

**Comment by alexellis** [comment]
Did you run this from the [troubleshooting guide - i.e. find-out-if-a-function-failed-to-start](https://docs.openfaas.com/deployment/troubleshooting/#find-out-if-a-function-failed-to-start)?

The command: `docker service ps --no-trunc myfunc` will reveal any errors with the Swarm service.

It sounds like your keychain or local login may be incorrect.

Do you have a script for setting up Harbor or a test registry someone could try?

**Comment by ziggekatten** [comment]
Yes, I have done docker login on my PC. To sum up
:
- Running "faas-cli up -a" builds, pushes and deploy the fucntion in the cluster, and it shows up in docker swarm. 

- Doing an "docker service ls" verifies this. Doing a "docker service logs <function name>" shows the error "image not found"

- Changing my docker repo (we use VMware Harbor) from private to public everything works 

- My config.json is empty, in a sense that it contains only "{}"

**Comment by alexellis** [comment]
Have you tried running `docker login harbor.coreteam.nu`? The config file should definitely not be empty, that sounds like an issue with your local Docker installation.

**Comment by ziggekatten** [comment]
Yep, I've been digging into this a bit now. First I did a "docker logout" for both docker hub and for harbor, clearing my credentials.

Now, doing a docker login SHOULD update the config.json. But it dows not. Searchi around, I found out, at when using Docker Desktop for Windows, the config.json is stored at c:\Users\<username>\.docker\config.json

The file content:
`{
	"auths": {
		"harbor.coreteam.nu": {}
	},
	"HttpHeaders": {
		"User-Agent": "Docker-Client/19.03.2 (windows)"
	},
	"credsStore": "desktop",
	"experimental": "enabled",
	"stackOrchestrator": "swarm"
}`
Trying to copy this to the config.json in the path that faas-cli like it does not help

**Comment by alexellis** [comment]
If we run Harbor we could try it out, but this may be more involved than the community can support on goodwill. 

My question for you would be, can you deploy any swarm services at all using the docker CLI with Auth turned on for the repo? If the answer is no, then fix that first then come back to us.

**Comment by ziggekatten** [comment]
All creds to rgee0!

It works like a charm now!

**Comment by ziggekatten** [comment]
The setting of the DOCKER_CONFIG solved it. Copying the config.json to the directory where faas-cli was looking (H:\.docker) did NOT work. ALmost ssems that if faas-cli does not find the DOCKER_CONFIG env variable it does not read the file even if it is there. I'll test that hypotheses at some point.

Normal docker pull/push have been working for ages with this repo.

**Comment by rgee0** [comment]
> Almost seems that if faas-cli does not find the DOCKER_CONFIG env variable it does not read the file even if it is there. I'll test that hypotheses at some point.

No need, I can tell you that it DOCKER_CONFIG is set then it will use that value to access the config file.  If it isn't set it will revert to a default which hangs off ~/.  For some reason your home is set to `H:` when I'd expect it to be somewhere under `c:\Users\%USERNAME%\`.  Maybe this is something you've consciously changed?

**Comment by rgee0** [comment]
@alexellis can you move this over to the CLI, please?  I had the option over on faas but don't get it on this repo.

**Comment by alexellis** [comment]
I think I've spent about an hour commenting on this, another an hour researching it and two more hours trying to reproduce the issue tonight. I can confirm that Swarm, OpenFaaS and MacOS all work fine with Harbor. I used Harbor 1.9 hosted on Linux Ubuntu. https://github.com/goharbor/harbor/releases/tag/v1.9.0

<img width="1197" alt="Screenshot 2019-10-11 at 21 27 18" src="https://user-images.githubusercontent.com/6358735/66682843-f8b96200-ec6d-11e9-809a-c7ec4c55b7a1.png">

<img width="1293" alt="Screenshot 2019-10-11 at 21 27 25" src="https://user-images.githubusercontent.com/6358735/66682859-fe16ac80-ec6d-11e9-9007-da7e6a07937b.png">

<img width="1786" alt="Screenshot 2019-10-11 at 21 27 09" src="https://user-images.githubusercontent.com/6358735/66682866-01119d00-ec6e-11e9-9d32-1f686f12e088.png">

**Comment by alexellis** [comment]
You are a collaborator now @rgee0, if you want to move it, feel free, but I can't reproduce the issue as per above.

**Comment by rgee0** [comment]
The issue seems to be in the user's environment.  I'll move it to faas-cli before closing so its easier to find for future searches / referencing, etc.
